ssl/statem/statem.c

```diff
@@ -849,10 +849,24 @@ static SUB_STATE_RETURN write_state_machine(SSL *s)
                 SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
                 return SUB_STATE_ERROR;
             }
-            if (confunc != NULL && !confunc(s, &pkt)) {
-                WPACKET_cleanup(&pkt);
-                check_fatal(s);
-                return SUB_STATE_ERROR;
+            if (confunc != NULL) {
+                int tmpret;
+
+                tmpret = confunc(s, &pkt);
+                if (tmpret <= 0) {
+                    WPACKET_cleanup(&pkt);
+                    check_fatal(s);
+                    return SUB_STATE_ERROR;
+                } else if (tmpret == 2) {
```

Comment:
Could we at least have internal macros for these constants (2, ...)?
Is `statem.h` a public header?  If not, introduction of a purely internal enum does not introduce any API issues I can think of.  You can still use the same enum from master I think, but otherwise equivalent internal macros.

Reply:
Note that the enum does not have to appear in function (return value) signatures, those can remain int-valued.  The enum is then just a `better` CPP macro.  There'll be no warnings.  The below code is valid C, whatever the storage type of the enum is, it gets automatically promoted to an int as needed.

```
#include <stdio.h>

enum foo {
    FOO = 42
};

int foo(void)
{
    return FOO;
}

int main(int argc, char *argv[])
{
    printf("%d\n", foo());
    return 0;
}
```

Reply:
Even if we do not change the return value signatures, this would still touch a lot of code and is essentially a refactoring, i.e. everywhere we have a "return" statement in any construction function would have to change to use the macros. I do not think that kind of change is appropriate for 3.0 (i.e. it does not meet the requirements of the [stable release updates policy](https://www.openssl.org/policies/technical/stable-release-updates.html))

Reply:
That's a tenable position, but one might also note that making all those point changes could make it easier to backport future PRs that overlap the code in question.  Your call I guess.  If you still think this is the best way forward, I'll continue the review as-is...

Reply:
It's a tricky call (both ways have good arguments), but I'd still call it on the side of changing less code in the 3.0 branch - so I still think the way in this PR is the best way forward.